Added /reports/coupons and /reports/coupons/stats endpoints

peterfabian opened the pull request:
Fixes #845 
Closes #861 

TODO:
- [x] tests
- [x] Add extended info as requested in #861
- [x] Initial generating info & confirm it's plugged into `Rebuild reports` -> moved into https://github.com/woocommerce/wc-admin/pull/1050

### Screenshots

### Detailed test instructions:

- Test both endpoints, ideally extensively

<!--- Note: When displaying information based on sample data, such as SwaggerHub, 
be sure to detail parts affected in Release Notes --->

joshuatf commented on includes/api/class-wc-admin-rest-reports-coupons-stats-controller.php:
```diff
@@ -291,7 +290,7 @@ public function get_collection_params() {
+		$params['code']  = array(
 			'description'       => __( 'Limit result set to items assigned one or more code.', 'wc-admin' ),
 			'type'              => 'array',
 			'sanitize_callback' => 'wp_parse_slug_list',
```
Do you think we should use the coupon ID here since we can reference that directly in the table?  Or are slugs a better approach here?

peterfabian replied:
Not sure, implemented it based on swagger, we can switch easily if needed.

joshuatf replied:
I'm partial towards using the ID here if possible since it will be consistent with the other endpoints and might make debugging a bit easier if needed in the future.  But I'll leave this up to your judgement.

joshuatf commented on includes/data-stores/class-wc-admin-reports-coupons-data-store.php:
```diff
@@ -0,0 +1,253 @@
+	 * @param array $coupon_data Coupon data.
+	 * @param array $query_args Query parameters.
+	 */
+	protected function include_extended_info( &$coupon_data, $query_args ) {
```
Nice! 👍

peterfabian commented on includes/api/class-wc-admin-rest-reports-coupons-controller.php:
```diff
@@ -104,7 +100,7 @@ public function get_items( $request ) {
 	 */
 	public function prepare_item_for_response( $report, $request ) {
-		$data = get_object_vars( $report );
+		$data = $report;
```
It's done this way because the filter further down needs the original `$report` object, while other functions in between can change the response data.

joshuatf replied:
Ah, I see it now!  My mistake.

peterfabian commented on includes/api/class-wc-admin-rest-reports-coupons-controller.php:
```diff
@@ -171,6 +167,45 @@ public function get_item_schema() {
+						'description' => __( 'Coupon code.', 'wc-admin' ),
+					),
+					'date_created'     => array(
+						'type'        => 'string',
```
You're right, fixed.

peterfabian commented on includes/data-stores/class-wc-admin-reports-coupons-data-store.php:
```diff
@@ -0,0 +1,253 @@
+	 */
+	protected $report_columns = array(
+		'coupon_id'      => 'coupon_id',
+		'gross_discount' => 'SUM(coupon_gross_discount) as gross_discount',
```
Correct, fixed.

peterfabian commented on includes/data-stores/class-wc-admin-reports-coupons-stats-data-store.php:
```diff
@@ -0,0 +1,216 @@
+		);
+		$query_args = wp_parse_args( $query_args, $defaults );
+
+		$cache_key = $this->get_cache_key( $query_args );
```
Yup, good catch.

peterfabian commented on includes/data-stores/class-wc-admin-reports-orders-data-store.php:
```diff
@@ -394,7 +329,7 @@ public function get_data( $query_args ) {
 
-			if ( $this->intervals_missing( $expected_interval_count, $db_interval_count, $intervals_query['per_page'], $query_args['page'], $query_args['order'], $query_args['orderby'], count( $intervals ) ) ) {
+			if ( WC_Admin_Reports_Interval::intervals_missing( $expected_interval_count, $db_interval_count, $intervals_query['per_page'], $query_args['page'], $query_args['order'], $query_args['orderby'], count( $intervals ) ) ) {
 				$this->fill_in_missing_intervals( $db_intervals, $query_args['adj_after'], $query_args['adj_before'], $query_args['interval'], $data );
```
Yes, exactly. Are you thinking of a better name?

joshuatf replied:
No, I think the name is great actually 👍 Just double checking to make sure I understood correctly.

peterfabian commented:
Hi @joshuatf. Thanks for review! I believe I addressed all of the points. 

Regarding the `before`/`after`--I think all endpoints have 1 week as default, but maybe we need to revisit this and change the defaults. In which case, I guess, it should be changed for all endpoints, so haven't changed it here.

joshuatf commented:
@peterfabian Changes look great!  One comment on the IDs vs slugs, but this is your call and pre-approving either way.

>Regarding the before/after--I think all endpoints have 1 week as default, but maybe we need to revisit this and change the defaults. In which case, I guess, it should be changed for all endpoints, so haven't changed it here.

Agreed - definitely outside the scope of this PR.  Opened a separate issue [here](https://github.com/woocommerce/wc-admin/issues/1103) and we can decide whether or not to move forward with changing those defaults.

joshuatf commented:
@peterfabian Gave this a rebase for you.  Could you take a look at the ID/slug item I mentioned above and push a change for that or go ahead and merge as is if you think slugs are a better option?

peterfabian commented:
@joshuatf 
Switched from using coupon codes to coupon ids in the filter, query parameter changed from `code` back to `coupons`, as it's no longer filtering by code.
